libraries/botbuilder-azure/src/cosmosDbPartitionedStorage.ts

```diff
@@ -0,0 +1,255 @@
+        if (!cosmosDbStorageOptions.databaseId) { throw new ReferenceError('databaseId is for CosmosDB required.'); }
+        if (!cosmosDbStorageOptions.containerId) { throw new ReferenceError('containerId for CosmosDB is required.'); }
+
+        this.cosmosDbStorageOptions = cosmosDbStorageOptions;
+    }
+
+    public async read(keys: string[]): Promise<StoreItems> {
+        if (!keys) { throw new ReferenceError(`Keys are required when reading.`); }
+        else if (keys.length === 0) { return {}; }
+
+        await this.initialize();
+
+        const storeItems: StoreItems = {};
+
+        await Promise.all(keys.map(async (k: string): Promise<void> => {
+            try {
+                const escapedKey = CosmosDbKeyEscape.escapeKey(k);
+
+                const readItemResponse = await this.container.item(escapedKey, escapedKey).read<DocumentStoreItem>();
+                const documentStoreItem = readItemResponse.resource;
+                if (documentStoreItem) {
+                    storeItems[documentStoreItem.realId] = documentStoreItem.document;
+                    storeItems[documentStoreItem.realId].eTag = documentStoreItem._etag;
+                }
+            } catch (err) {
+                // When an item is not found a CosmosException is thrown, but we want to
+                // return an empty collection so in this instance we catch and do not rethrow.
+                // Throw for any other exception.
+                if (err.code === 404) { }
+                // Throw unique error for 400s
+                else if (err.code === 400) {
+                    this.throwInformativeError(`Error initializing container. You might be using partitions in a non-partitioned DB or
+                    are not using partitions in a partitioned db that already contains partitioned data`, err);
+                } else {
+                    this.throwInformativeError('Error reading from container', err);
+                }
+            }
+        }));
+
+        return storeItems;
+    }
+
+    public async write(changes: StoreItems): Promise<void> {
+        if (!changes) { throw new ReferenceError(`Changes are required when writing.`); }
+        else if (changes.length === 0) { return; }
+
+        await this.initialize();
+
+        await Promise.all(Object.keys(changes).map(async (k: string): Promise<void> => {
+            const changesCopy: any = {...changes[k]};
+
+            // Remove eTag from JSON object that was copied from IStoreItem.
+            // The ETag information is updated as an _etag attribute in the document metadata.
+            delete changesCopy.eTag;
+            const documentChange = new DocumentStoreItem({
+                id: CosmosDbKeyEscape.escapeKey(k),
+                realId: k,
+                document: changesCopy
+            });
+
+            const eTag: string = changes[k].eTag;
+            if (!eTag || eTag === '*') {
```

Comment:
I don't think you need to differentiate upsert vs replace on etag here. Upsert IfMatch will always succeed if it doesn't exist.

I wrote up a simple proof of concept showing this off.
https://gist.github.com/christopheranderson/57081bce1236a6941a553e9b9333c22d

Reply:
Can you clarify that for me? I think I get it but want to make sure before we merge...I should get rid of `replace` and just use `upsert` for both `if` branches, right?

Reply:
@christopheranderson Hey Chris, do you have anything to add for these two unresolved conversations, or would you say we're good to merge?

Reply:
Yes, you can just use upsert. You can probably avoid the if condition and use a ternary statement instead, as well.

Reply:
Your code works fine, so feel free to resolve if you want, but you could shrink your LOC count down. :)

Reply:
good call. done.